### ISSUE TITLE
Add a `Secret` type to the Node.js SDK which can be serialized into encrypted parameters instead of source

### ISSUE THREAD

**Issue by lukehoban** [pulumi/pulumi#397]
When we serialize closures, we capture by value all referenced objects/variables and emit them as JavaScript text into the body of a Lambda (or in the future container).  

We would like to be able to identify sensitive values, and transfer these into the runtime environment in an encrypted form.  For example, using Lambda's encrypted environment parameters (which are implicitly decrypted by the Lambda KMS key inside the Lambda function runtime environment - but ensure that the variable is never passed in plain text).

I imagine we would stamp this type with a special Symbol such that our closure serialization can know when it's hit one of these and serialize into a different form of storage than the standard JSON-like serialization.

This `Secret` wrapper object may be the same one we use as the type of 1st class "secrets" in the config system if we end up having such a thing.

Related to #368 and #35.

**Comment by joeduffy** [comment]
Given all the open questions around secret management generally -- a very important area, but one we want to be deliberate and planful about -- do we really think this belongs in M16?

**Comment by lukehoban** [comment]
Moving out, but would love to start having the design discussions here (and ideally a strawman proposal) so that we can make sure we know what approach(es) we will take here.

**Comment by joeduffy** [comment]
I believe we have a design for how to accomplish this. We'll start making progress during M20. This particular feature is likely to need to wait until the foundation is laid (so M21 or thereabouts), but I'll keep it in M20 for now so we can link to it and discuss as part of the overall design discussion.

**Comment by lukehoban** [comment]
@pgavlin Could you summarize (or link to) current plan for this?

**Comment by pgavlin** [comment]
After some offline discussion, we've elected to move forwards with "Secrets as a flag on `Output<T>`". This approach is very straightforward, has well-understood behavior, and composes nicely.

The two approaches that would have added a `Secret` type suffered from practical issues regarding composition:

- Potentially-secret outputs would have needed to be `Output<T | Secret<T>>`
- `Input<T>` would have needed to accept `Secret<T>` and `Output<Secret<T>>`

Because components generally operate on `Input<T>` values, they would have been required to explicitly handle both secret and non-secret inputs. As secret inputs would have had their own type, this would have resulted in a large amount of duplicated code.

We agreed that the primary goal of this work is to improve the privacy and security of persisted values. We are therefore tentatively comfortable with not exposing the secret-ness of a value through the type system.

The behaviors of the `Output.apply`, `pulumi.all`, and `pulumi.output` methods deserve some discussion.

1. The result of `Output.apply` on a secret output is itself secret
2. The result of `pulumi.all` is secret iff any of the inputs to the `all` is a secret output
3. The result of `pulumi.output` is secret iff any value reachable from the root value is a secret output

The final behavior prompted some worry about the potential virality of secret-ness. Consider a large object with a single, deeply nested secret value:

```
const myObj = {
    foo: {
        bar: {
            baz: {
                qux: // some secret value
            },
        },
    },
};
```
If this object is lifted using `pulumi.output`, the entire result will be treated as secret. We believe that we may be able to do better in this case up to the point at which the value is transformed by an apply (at which time we lose visibility into the precise characteristics of the result).

**Comment by lukehoban** [comment]
As part of delivering this, we will ensure that secrets are stored in encrypted form in the checkpoint file (previously tracked in pulumi/pulumi#2034).

**Comment by joeduffy** [comment]
These are out for review, but I am pushing to M23 so we can close the prior sprint.

**Comment by ellismg** [comment]
Support for this has landed in 0.17.11.
